include/mapnik/geometry_is_valid.hpp

```diff
@@ -106,6 +110,140 @@ inline bool is_valid(T const& geom)
     return detail::geometry_is_valid() (geom);
 }
 
+typedef boost::geometry::model::d2::point_xy<double> Point;
+typedef boost::geometry::model::segment<Point> Segment;
+
+template <typename T>
+std::vector<Segment> make_segment_vector(linear_ring<T> ring)
+{
+    std::vector<Segment> result;
+
+    if (ring.size() > 1) {
+        for (int i = 0; i < ring.size()-1; ++i)
```

Comment:
I recommend caching the value of ring.size() (just in case the compiler is not smart enough to do this) and calling `result.reserve` since the final size of the `result` is known. Should lead to small speedups.

```cpp
std::size_t num_rings = ring.size();
result.reserve(num_rings-1);
```